src/main/java/core/basesyntax/SalaryInfo.java

```diff
@@ -39,8 +46,83 @@
      * Андрей - 600
      * София - 900</p>
      */
-    public String getSalaryInfo(String[] names, String[] data, String dateFrom, String dateTo)
-            throws Exception {
-        return null;
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("dd.MM.yyyy");
+
+    public String getSalaryInfo(String[] names, String[] data, String dateFrom, String dateTo) {
+        LocalDate from = parseDate(dateFrom);
+        LocalDate to = parseDate(dateTo);
+        if (from.isAfter(to)) {
+            throw new IllegalDateParametersException("Wrong parameters");
+        }
+
+        Map<String, Integer> nameToEarnings = createNameToEarningsMap(names);
+
+        for (String record : data) {
+            EarningsRecord earningsRecord = EarningsRecord.create(record.split(" "));
+            if (filterEarningsRecord(earningsRecord, from, to, nameToEarnings)) {
+                processRecord(earningsRecord, nameToEarnings);
+            }
+        }
+        return generateReport(names, dateFrom, dateTo, nameToEarnings);
+    }
+
+    private String generateReport(String[] names, String dateFrom, String dateTo,
+            Map<String, Integer> nameToEarnings) {
+        StringBuilder sb = new StringBuilder("Отчёт за период " + dateFrom + " - " + dateTo);
+        for (String name : names) {
+            sb.append("\n").append(name).append(" - ").append(nameToEarnings.get(name));
+        }
+        return sb.toString();
+    }
+
+    private Map<String, Integer> createNameToEarningsMap(String[] names) {
+        Map<String, Integer> nameToEarnings = new HashMap<>();
+        for (String name : names) {
+            nameToEarnings.put(name, 0);
+        }
+        return nameToEarnings;
+    }
+
+    private boolean filterEarningsRecord(
+            EarningsRecord earningsRecord,
+            LocalDate fromDate,
+            LocalDate toDate,
+            Map<String, Integer> nameToEarnings
+    ) {
+        return nameToEarnings.containsKey(earningsRecord.name)
+                && !earningsRecord.date.isBefore(fromDate)
+                && !earningsRecord.date.isAfter(toDate);
     }
+
+    private void processRecord(EarningsRecord earningsRecord, Map<String, Integer> nameToEarnings) {
+        Integer earnings = nameToEarnings.get(earningsRecord.name)
+                + earningsRecord.hours * earningsRecord.hourlyRate;
+        nameToEarnings.put(earningsRecord.name, earnings);
+    }
+
+    private static LocalDate parseDate(String date) {
```

Comment:
static is evil :) don't create static methods

Reply:
solved